src/build.d

```diff
@@ -1480,6 +1486,76 @@ class Dependency
         }
     }
 
+     // Metadata required for parallel execution
+    private
+    {
+        shared uint builtDeps; /// amount of dependencies already built
+        bool visited; /// whether this dependency was already visited
+        Dependency[] requiredBy; /// dependencies relying on this one
+    }
+
+    /**
+    Builds all supplied targets using the global taskPool to execute buildRecursive
+    for every leaf dependency.
+
+    Params:
+        targets = dependencies to build
+    **/
+    static void build(Dependency[] targets)
+    {
+        import core.atomic: atomicLoad, atomicOp, atomicStore;
+
+        //// Recursively appends all dependencies without further dependencies to leafs
+        static void findLeafs(Dependency cur, ref Dependency[] leafs)
+        {
+            if (cur.visited) return;
+            cur.visited = true;
+
+            if (cur.deps.empty)
+            {
+                leafs ~= cur;
+            }
+            else
+            {
+                foreach (dep; cur.deps)
+                {
+                    dep.requiredBy ~= cur;
+                    findLeafs(dep, leafs);
+                }
+            }
+        }
+
+        /// Recursively builds cur and other dependencies requiring it
+        /// (if cur is the last missing dependency)
+        static void buildRecursive(Dependency cur)
+        {
+            // Exit if any dependency failed to build
+            if (atomicLoad(aborted)) return;
+
+            cur.run();
+
+            // Mark this dependency as finished for all parent dependencies
+            auto unblocked = cur.requiredBy
+                .filter!(p => p.builtDeps.atomicOp!"+="(1) == p.deps.length)
+                .array; // lazy cache produces segfault in parallel
+
+            // Execute all unblocked dependencies
+            taskPool.parallel(unblocked, 1).each!buildRecursive;
```

Comment:
It looks like this still isn't fully parallel.  The test case I sent you demonstrates that only one "leg" of the tree gets executed in parallel at a time.